File: c++/modifyingSequenceTest.cpp
#include "test.h"
#include "modifyingSequence.hpp"

using namespace std;

class ModifyingSequence : public ::testing::Test
{
protected:
    virtual void SetUp()
    {
        v2.resize(v.size());
    }
    vector<int> v{10, 5, 4, 3, 20, 190};
    vector<int> v2;
};

TEST_F(ModifyingSequence, copys)
{
    my::copy(v.begin(), v.end(), v2.begin());
    ASSERT_EQ(v[0], v2[0]);

    v2.clear();
    my::copy_if(v.begin(), v.end(), back_inserter(v2), 
            [](int x){ return x > 5; });
    ASSERT_EQ(3, v2.size());
    ASSERT_EQ(v[0], v2[0]);

    v2.clear();
    my::copy_n(v.begin(), 2, back_inserter(v2));
    ASSERT_EQ(2, v2.size());
    ASSERT_EQ(v[1], v2[1]);
    ASSERT_EQ(v[0], v2[0]);

    v2.resize(v.size());
    my::copy_backward(v.begin(), v.end(), v2.end());
    ASSERT_EQ(v, v2);
}

TEST_F(ModifyingSequence, moves)
{
    my::move(v.begin(), v.end(), v2.begin());
    ASSERT_EQ(v[0], v2[0]);

    --v2[0];
    my::move_backward(v.begin(), v.end(), v2.end());
    ASSERT_EQ(v, v2);
}

TEST_F(ModifyingSequence, fills)
{
    my::fill(v2.begin(), v2.end(), 1);
    ASSERT_EQ(1, v2[0]);
    ASSERT_EQ(1, *prev(v2.end()));

    my::fill_n(v2.begin(), 2, 2);
    ASSERT_EQ(2, v2[0]);
    ASSERT_EQ(2, v2[1]);
}

TEST_F(ModifyingSequence, transform)
{
    my::transform(v.begin(), v.end(), v.begin(), [](int &x){ return ++x; });
    vector<int> vectorplusOne{11, 6, 5, 4, 21, 191};
    ASSERT_EQ(vectorplusOne, v);
}

TEST_F(ModifyingSequence, generates)
{
    my::generate(v2.begin(), v2.end(), [](){ return 1; });
    ASSERT_EQ(1, v2[0]);
    ASSERT_EQ(1, *prev(v2.end()));

    my::generate_n(v2.begin(), 2, [](){ return 2; });
    ASSERT_EQ(2, v2[0]);
    ASSERT_EQ(2, v2[1]);
}

TEST_F(ModifyingSequence, removes)
{
    v.erase(my::remove(v.begin(), v.end(), 5), v.end());
    ASSERT_EQ(5, v.size());

    v.erase(my::remove_if(v.begin(), v.end(), 
                [](int x){ return x > 10; }), v.end());
    ASSERT_EQ(3, v.size());

    v2.clear();
    my::remove_copy(v.begin(), v.end(), back_inserter(v2), 10);
    ASSERT_EQ(2, v2.size());
    ASSERT_EQ(v2.end(), find(v2.begin(), v2.end(), 10));

    my::remove_copy_if(v.begin(), v.end(), back_inserter(v2), 
            [](int x){ return x != 10; });
    ASSERT_EQ(3, v2.size());
    ASSERT_EQ(prev(v2.end()), find(v2.begin(), v2.end(), 10));
}

TEST_F(ModifyingSequence, replaces)
{
    my::replace(v.begin(), v.end(), 10, 11);
    ASSERT_EQ(11, v[0]);
    my::replace(v.begin(), v.end(), 190, 10);
    ASSERT_EQ(10, *prev(v.end()));

    my::replace_if(v.begin(), v.end(), [](int x){ return x > 5; }, 5);
    const int fives = count(v.begin(), v.end(), 5);
    ASSERT_EQ(4, fives);
}

TEST_F(ModifyingSequence, replace_copys)
{
    my::replace_copy(v.begin(), v.end(), v2.begin(), 5, 6);
    ASSERT_EQ(v2.end(), find(v2.begin(), v2.end(), 5));
    ASSERT_EQ(++v2.begin(), find(v2.begin(), v2.end(), 6));

    v2.clear();
    my::replace_copy_if(v.begin(), v.end(), back_inserter(v2),
            [](int x){ return x > 10; }, 10);
    const int tens = count(v2.begin(), v2.end(), 10);
    ASSERT_EQ(3, tens);
    ASSERT_EQ(v2.end(), find(v2.begin(), v2.end(), 190));
}

TEST_F(ModifyingSequence, swaps)
{
    vector<int> replicaV(v);
    vector<int> replicaV2(v2);
    my::swap(v, v2);
    ASSERT_EQ(replicaV, v2);
    ASSERT_EQ(replicaV2, v);

    // make sure v.size() == v2.size()
    my::swap_ranges(v.begin(), v.end(), v2.begin());
    ASSERT_EQ(replicaV, v);
    ASSERT_EQ(replicaV2, v2);

    auto vIt = v.begin();
    auto v2It = v2.begin();
    while (vIt != v.end()) {
        my::iter_swap(vIt++, v2It++);
    }
    ASSERT_EQ(replicaV2, v);
    ASSERT_EQ(replicaV, v2);
}

TEST_F(ModifyingSequence, reverse)
{
    vector<int> reversed{190, 20, 3, 4, 5, 10};
    my::reverse(v.begin(), v.end());
    ASSERT_EQ(reversed, v);
    reverse(reversed.begin(), reversed.end());
    my::reverse_copy(v.begin(), v.end(), v2.begin());
    ASSERT_EQ(reversed, v2);
}

TEST_F(ModifyingSequence, rotates)
{
    int ary[5] = { 1, 2, 3, 4, 5 };
    vector<int> rotatedAry{ 3, 4, 5, 1, 2 };
    vector<int> rotatedAry2{1, 2, 3, 4, 5 };
    my::rotate(ary, ary + 2, ary + 5);
    vector<int> tmp(ary, ary + 5);
    my::rotate(ary, ary + 3, ary + 5);
    vector<int> tmp2(ary, ary + 5);

    ASSERT_EQ(rotatedAry, tmp);
    ASSERT_EQ(rotatedAry2, tmp2);

    auto pivot = next(v.begin(), 3);
    my::rotate_copy(v.begin(), pivot, v.end(), v2.begin());
    vector<int> rotatedV{ 3, 20, 190, 10, 5, 4 };
    ASSERT_EQ(rotatedV, v2);
}

TEST_F(ModifyingSequence, shuffle)
{
    vector<int> replicaV(v);
    my::shuffle(v.begin(), v.end());
    ASSERT_TRUE(is_permutation(v.begin(), v.end(), replicaV.begin()));
}

TEST_F(ModifyingSequence, uniques)
{
    unsigned int size = v.size();
    // there is not duplicate value in v before
    v.push_back(v.back());
    v.erase(my::unique(v.begin(), v.end()), v.end());
    ASSERT_EQ(size, v.size());

    vector<int> duplicate{ 1, 1, 2, 2, 3 };
    vector<int> uniqueness{ 1, 2, 3 };
    vector<int> tmp;
    my::unique_copy(duplicate.begin(), duplicate.end(), 
            back_inserter(tmp));
    ASSERT_EQ(uniqueness, tmp);
}

File: php/ModifyingSequenceTest.php
<?php
namespace Algorithm;

class ModifyingSequenceTest extends \PHPUnit_Framework_TestCase
{
    private $ary;

    /*
     * @before
     */
    public function setUp() {
        $this->ary = [1, 2, 3, 4, 6];
    }

    public function testCopy()
    {
        $this->assertEquals([1, 2, 3], copy($this->ary, 0, 3));
    }

    public function testCopyIf()
    {
        $this->assertEquals([1, 3], copyIf($this->ary, function($x) {
            return $x % 2 === 1;
        }));
    }

    public function testCopyN()
    {
        $ary = [0, 0, 0, 1];
        $this->assertEquals([3, 3, 3, 1], copyN($ary, 3, 3));
        $ary2 = [0, 0];
        $this->assertEquals([3, 3], copyN($ary2, 3, 3));
    }

    public function testCopyBackward()
    {
        $ary = [2, 2, 2];
        copyBackward($ary, $this->ary);
        $this->assertEquals([3, 4, 6], $ary);
        $ary = array_pad([], 6, 1);
        copyBackward($ary, $this->ary);
        $this->assertEquals([1, 1, 2, 3, 4, 6], $ary);
    }

    public function testFill()
    {
        $this->assertEquals([], fill([], 3));
        $this->assertEquals([3, 3], fill([1, 1], 3));
    }

    public function testFillN()
    {
        $this->assertEquals([3, 3, 3], fillN([], 3, 3));
        $this->assertEquals([3, 3, 3], fillN([2, 2], 3, 3));
        $this->assertEquals([3, 3, 2], fillN([2, 2, 2], 2, 3));
    }

    public function testTransform()
    {
        $func = function($x) {
            return $x + 1;
        };
        $this->assertEquals(array_map($func, $this->ary),
            transform($this->ary, $func));
    }

    public function testGenerate()
    {
        $func = function() {
            return 2;
        };
        $this->assertEquals([2, 2, 2], generate([1, 1, 1], $func));
    }

    public function testGenerateN()
    {
        $func = function() {
            return 1;
        };
        $this->assertEquals([1, 1], generateN([0, 0], 2, $func));
        $this->assertEquals([1, 1], generateN([0, 0], 3, $func));
    }

    public function testRemove()
    {
        $ary = [1, 2, 3];
        remove($ary, 3);
        $this->assertEquals([1, 2], $ary);
    }

    public function testRemoveIf()
    {
        $ary = [1, 2, 3];
        removeIf($ary, function($e) {
            return $e < 2;
        });
        $this->assertEquals([2, 3], $ary);
    }

    public function testRemoveCopy()
    {
        $ary = [1, 2, 3];
        $this->assertEquals([3], removeCopy($ary, 3));
        $this->assertEquals([1, 2], $ary);
    }

    public function testRemoveCopyIf()
    {
        $ary = [1, 2, 3];
        $this->assertEquals([1], removeCopyIf($ary, function($e) {
            return $e < 2;
        }));
        $this->assertEquals([2, 3], $ary);
    }

    public function testReplace()
    {
        $ary = [1, 3, 3];
        replace($ary, 3, 2);
        $this->assertEquals([1, 2, 2], $ary);
    }

    public function testReplaceIf()
    {
        $ary = [1, 2, 3, 3];
        replaceIf($ary, function($e) {
            return $e >= 2;
        }, 1);
        $this->assertEquals([1, 1, 1, 1], $ary);
    }

    public function testReplaceCopy()
    {
        $ary = [1, 3, 3];
        $this->assertEquals([1, 2, 2], replaceCopy($ary, 3, 2));
    }

    public function testReplaceCopyIf()
    {
        $ary = [1, 2, 3, 3];
        $this->assertEquals([1, 1, 1, 1], replaceCopyIf($ary, function($e) {
            return $e >= 2;
        }, 1));
    }

    public function testSwap()
    {
        $a = 1;
        $b = 2;
        swap($a, $b);
        $this->assertEquals(2, $a);
        $this->assertEquals(1, $b);
    }

    public function testSwapRange()
    {
        $a = [1, 2, 3, 4];
        $b = [4, 3, 2, 1];
        swapRange($a, $b, 1, 3);
        $this->assertEquals([1, 3, 2, 4], $a);
        $this->assertEquals([4, 2, 3, 1], $b);

        // out of range
        $a = [1, 1, 1];
        $b = [2];
        swapRange($a, $b, 1, 3);
        $this->assertEquals([1, 1, 1], $a);
        $this->assertEquals([2], $b);

        // some part out of range
        $c = [2, 2];
        swapRange($a, $c, 1, 3);
        $this->assertEquals([1, 2, 1], $a);
        $this->assertEquals([2, 1], $c);
    }

    public function testReverse()
    {
        $ary = [1, 2, 3];
        reverse($ary);
        $this->assertEquals([3, 2, 1], $ary);
        $ary2 = [1, 2, 3, 4];
        reverse($ary2);
        $this->assertEquals([4, 3, 2, 1], $ary2);
    }

    public function testReverseCopy()
    {
        $ary = [1, 2, 3];
        $this->assertEquals([3, 2, 1], reverseCopy($ary));
    }

    public function testShuffle()
    {
        shuffle($this->ary);
    }

    public function testRotate()
    {
        $ary = [1, 2, 3, 4];
        rotate($ary, 2);
        $this->assertEquals([3, 4, 1, 2], $ary);

        rotate($ary, 5);
        $this->assertEquals([4, 1, 2, 3], $ary);

        rotate($ary, -1);
        $this->assertEquals([3, 4, 1, 2], $ary);
    }

    public function testRotateCopy()
    {
        $this->assertEquals([], rotateCopy([], 3));
        $this->assertEquals([3, 4, 6, 1, 2], rotateCopy($this->ary, 2));
        $this->assertEquals([3, 4, 6, 1, 2], rotateCopy([1, 2, 3, 4, 6], -3));
        $this->assertEquals([3, 4, 6, 1, 2], rotateCopy([1, 2, 3, 4, 6], 7));
    }

    public function testUnique()
    {
        $dup = [1, 1, 2, 2, 3];
        unique($dup);
        $this->assertEquals([1, 2, 3], $dup);
    }

    public function testUniqueCopy()
    {
        $dup = [1, 1, 2, 2, 3];
        $this->assertEquals([1, 2, 3], uniqueCopy($dup));
    }
}
?>
